node/src/components/chain_synchronizer/operations.rs

```diff
@@ -100,42 +131,54 @@ impl<'a> ChainSyncContext<'a> {
             filter_count: AtomicI64::new(0),
             locally_available_block_range_on_start,
             trie_fetch_limit: Semaphore::new(config.max_parallel_trie_fetches()),
-        }
-    }
+        };
 
-    /// Initializes the trusted block header.
-    ///
-    /// # Panics
-    ///
-    /// Will panic if set more than once.
-    fn set_trusted_block_header(&mut self, header: &'a BlockHeader) {
-        if self.trusted_block_header.is_some() {
-            panic!("cannot set trusted block header twice");
-        }
+        let trusted_block_header = match config.trusted_hash() {
+            Some(trusted_hash) => {
+                *fetch_and_store_initial_trusted_block_header(&ctx, metrics, trusted_hash).await?
+            }
+            None => match effect_builder.get_highest_block_header_from_storage().await {
+                Some(block_header) => block_header,
+                None => {
+                    debug!("no highest block header found in storage");
+                    return Ok(None);
+                }
+            },
+        };
+
+        ctx.trusted_block_header = Some(Arc::new(trusted_block_header));
```

Comment:
I'd have liked that, yes.  Unfortunately the context struct is passed down through more functions inside `fetch_and_store_initial_trusted_block_header` and this would need heavy refactoring to support your proposed approach.